zipkin/src/main/java/com/linecorp/armeria/common/tracing/RequestContextCurrentTraceContext.java

```diff
@@ -0,0 +1,160 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.internal.tracing.PingPongExtra;
+
+import brave.propagation.CurrentTraceContext;
+import brave.propagation.TraceContext;
+import io.netty.util.Attribute;
+import io.netty.util.AttributeKey;
+
+/**
+ * {@linkplain brave.Tracing.Builder#currentTraceContext(brave.propagation.CurrentTraceContext) Tracing
+ * context} implemented with {@linkplain com.linecorp.armeria.common.RequestContext}.
+ *
+ * <p>This {@link CurrentTraceContext} stores/loads the trace context into/from a
+ * {@link RequestContext}'s attribute so that there's no need for thread local variables
+ * which can lead to unpredictable behavior in asynchronous programming.
+ */
+public final class RequestContextCurrentTraceContext extends CurrentTraceContext {
+
+    public static final CurrentTraceContext INSTANCE = new RequestContextCurrentTraceContext();
+    private static final Logger logger = LoggerFactory.getLogger(RequestContextCurrentTraceContext.class);
+    private static final AttributeKey<TraceContext> TRACE_CONTEXT_KEY =
+            AttributeKey.newInstance("trace-context");
+
+    private static final Scope INCOMPLETE_CONFIGURATION_SCOPE = new Scope() {
+        @Override
+        public void close() {
+        }
+
+        @Override
+        public String toString() {
+            return "IncompleteConfigurationScope";
+        }
+    };
+
+    private static final Scope INITIAL_REQUEST_SCOPE = new Scope() {
+        @Override
+        public void close() {
+            // Don't remove the outer-most context (client or server request)
+        }
+
+        @Override
+        public String toString() {
+            return "InitialRequestScope";
+        }
+    };
+
+    /**
+     * Use this to ensure the trace context propagates to children.
+     *
+     * <p>Ex.
+     * <pre>{@code
+     *  // Ensure the trace context propagates to children
+     * ctx.onChild(RequestContextCurrentTraceContext::copy);
+     * }</pre>
+     */
+    public static void copy(RequestContext src, RequestContext dst) {
+        dst.attr(TRACE_CONTEXT_KEY).set(src.attr(TRACE_CONTEXT_KEY).get());
+    }
+
+    @Override
+    @Nullable
+    public TraceContext get() {
+        if (!ensureCurrentRequestContextOrWarnOnce()) {
+            return null;
+        }
+        return RequestContext.current().attr(TRACE_CONTEXT_KEY).get();
+    }
+
+    @Override
+    public Scope newScope(@Nullable TraceContext currentSpan) {
+        // Handle inspection added to ensure we can fail-fast if this isn't installed.
+        if (currentSpan != null && PingPongExtra.maybeSetPong(currentSpan)) {
+            return Scope.NOOP;
+        }
+
+        if (!ensureCurrentRequestContextOrWarnOnce()) {
+            return INCOMPLETE_CONFIGURATION_SCOPE;
+        }
+
+        final Attribute<TraceContext> contextAttribute = RequestContext.current().attr(TRACE_CONTEXT_KEY);
```

Comment:
It might be slightly better to use the result of `mapCurrent` instead of reading from the thread-local twice.

Reply:
very nice